src/content/en/updates/posts/2016/03/access-usb-devices-on-the-web.markdown

```diff
@@ -0,0 +1,370 @@
+configurations the device can have - basically all fields contained in the
+[device USB Descriptor](http://www.beyondlogic.org/usbnutshell/usb5.shtml#DeviceDescriptors)
+
+For info, if a USB device announces its [support for WebUSB](https://wicg.github.io/webusb/#webusb-platform-capability-descriptor),
+as well as defining a landing page URL, Chrome will show a persistent
+notification when the USB device is plugged in. Clicking on this notification
+will open the landing page.
+
+<img style="width:723px; max-height:345px" src="/web/updates/images/2016-03-02-access-usb-devices-on-the-web/web-usb-notification.png" alt="WebUSB notification"/>
+
+From there, you can simply call `navigator.usb.getDevices` and get access to
+your Arduino device as shown below.
+
+{% highlight javascript %}
+navigator.usb.getDevices().then(devices => {
+  devices.map(device => {
+    console.log(device.productName);      // "Arduino Micro"
+    console.log(device.manufacturerName); // "Arduino LLC"
+  });
+})
+{% endhighlight %}
+
+### Talk to an Arduino USB board
+
+Okay, now let's see how easy it is to communicate from a WebUSB compatible
+Arduino board over the USB port. Check out instructions at
+[https://github.com/webusb/arduino](https://github.com/webusb/arduino) to
+WebUSB-enable your [sketches](http://www.arduino.cc/en/Tutorial/Sketch).
+
+{% highlight javascript %}
+var device;
+
+navigator.usb.requestDevice({ filters: [{ vendorId: 0x2341 }] })
+.then(selectedDevice => {
+   device = selectedDevice;
+   return device.open(); // Begin a session.
+ })
+.then(() => device.selectConfiguration(1)) // Select configuration #1 for the device.
+.then(() => device.claimInterface(2)) // Request exclusive control over interface #2.
+.then(() => device.controlTransferOut({
+    requestType: 'class',
+    recipient: 'interface',
+    request: 0x22,
+    value: 0x01,
+    index: 0x02})) // Ready to receive data
+.then(() => device.transferIn(5, 64)) // Waiting for 64 bytes of data from endpoint #5.
+.then(result => {
+  let decoder = new TextDecoder();
+  console.log('Received: ' + decoder.decode(result.data));
+})
+.catch(error => { console.log(error); });
+{% endhighlight %}
+
+Please keep in mind that the WebUSB library we are using here is just
+implementing one example protocol (based on the standard USB serial protocol)
+and that you can create any set and types of endpoints you
+wish. Control transfers are especially nice for small configuration commands as
+they get bus priority and have a well defined structure.
+
+And here's the sketch that has been uploaded to the Arduino board.
+
+{% highlight cpp %}
+// Third-party WebUSB Arduino library
+#include <WebUSB.h>
+
+const WebUSBURL URLS[] = {
+  { 1, "webusb.github.io/arduino/demos/" },
+  { 0, "localhost:8000" },
+};
+
+const uint8_t ALLOWED_ORIGINS[] = { 1, 2 };
+
+WebUSB WebUSBSerial(URLS, 2, 1, ALLOWED_ORIGINS, 2);
+
+#define Serial WebUSBSerial
+
```

Comment:
This is not a section. `#define` is a C component that allows the programmer to give a name to a constant value before the program is compiled.